apps/src/applab/applab.js

```diff
@@ -1274,37 +1288,57 @@ function onDataViewChange(view, oldTableName, newTableName) {
   if (!getStore().getState().pageConstants.hasDataMode) {
     throw new Error('onDataViewChange triggered without data mode enabled');
   }
-  const storageRef = getProjectDatabase().child('storage');
+
+  const projectStorageRef = getProjectDatabase().child('storage');
+  const sharedStorageRef = getSharedDatabase().child('storage');
 
   // Unlisten from previous data view. This should not interfere with events listened to
   // by onRecordEvent, which listens for added/updated/deleted events, whereas we are
   // only unlistening from 'value' events here.
-  storageRef.child('keys').off('value');
-  storageRef.child(`tables/${oldTableName}/records`).off('value');
+  projectStorageRef.child('keys').off('value');
+  projectStorageRef.child(`tables/${oldTableName}/records`).off('value');
+  sharedStorageRef.child(`tables/${oldTableName}/records`).off('value');
   getColumnsRef(oldTableName).off();
 
   switch (view) {
     case DataView.PROPERTIES:
-      storageRef.child('keys').on('value', snapshot => {
+      projectStorageRef.child('keys').on('value', snapshot => {
         getStore().dispatch(updateKeyValueData(snapshot.val()));
       });
       return;
-    case DataView.TABLE:
-      // Add any columns which appear in records in Firebase to the list of columns in
-      // Firebase. Do NOT do this every time the records change, to avoid adding back
-      // a column shortly after it was explicitly renamed or deleted.
-      addMissingColumns(newTableName);
-
-      onColumnNames(newTableName, columnNames => {
-        getStore().dispatch(updateTableColumns(newTableName, columnNames));
-      });
-
-      storageRef
-        .child(`tables/${newTableName}/records`)
-        .on('value', snapshot => {
-          getStore().dispatch(updateTableRecords(newTableName, snapshot.val()));
+    case DataView.TABLE: {
+      if (
+        experiments.isEnabled(experiments.APPLAB_DATASETS) &&
+        getStore().getState().data.tableListMap[newTableName] ===
+          tableType.SHARED
+      ) {
```

Comment:
Instead of duplicating here, consider either factoring this into a helper function or trying something like this:
var storageRef;
```
if (experiment) {
  storageRef = sharedStorageRef
} else {
  ...
  storageRef = projectStorageRef
}
storageRef.child(...).on(...)
```